### PR TITLE
Increase a bunch of timeouts to reduce flakes

### LINKED ISSUE #14059: TestWatch is flaky
```console
	:29: generating index 1
	:29: creating /registry/controllers/default/foo, index 1 (ttl: 0)
	:29: returning /registry/controllers/default/foo: &etcd.Response{Action:"", Node:(*etcd.Node)(0xc2081bfb00), PrevNode:(*etcd.Node)(nil), EtcdIndex:0x0, RaftIndex:0x0, RaftTerm:0x0} 
	:29: returning /registry/controllers/default/foo: &etcd.Response{Action:"", Node:(*etcd.Node)(0xc2081bfb00), PrevNode:(*etcd.Node)(nil), EtcdIndex:0x0, RaftIndex:0x0, RaftTerm:0x0} 
	:29: generating index 2
	:29: updating /registry/controllers/default/foo, index 1 -> 2 (ttl: 0)
	:29: returning /registry/controllers/default/foo: &etcd.Response{Action:"", Node:(*etcd.Node)(0xc208092660), PrevNode:(*etcd.Node)(nil), EtcdIndex:0x0, RaftIndex:0x0, RaftTerm:0x0} 
	:29: returning /registry/controllers/default/foo: &etcd.Response{Action:"", Node:(*etcd.Node)(0xc208092660), PrevNode:(*etcd.Node)(nil), EtcdIndex:0x0, RaftIndex:0x0, RaftTerm:0x0} 
	:29: returning /registry/controllers/test: &etcd.Response{Action:"", Node:(*etcd.Node)(0xc208229680), PrevNode:(*etcd.Node)(nil), EtcdIndex:0x0, RaftIndex:0x0, RaftTerm:0x0} 
	:29: returning /registry/controllers/test: &etcd.Response{Action:"", Node:(*etcd.Node)(nil), PrevNode:(*etcd.Node)(nil), EtcdIndex:0x0, RaftIndex:0x0, RaftTerm:0x0} &etcd.EtcdError{ErrorCode:100, Message:"", Cause:"", Index:0x7b}
	:29: returning /registry/controllers/test: &etcd.Response{Action:"", Node:(*etcd.Node)(0xc208268060), PrevNode:(*etcd.Node)(nil), EtcdIndex:0x0, RaftIndex:0x0, RaftTerm:0x0} 
=== RUN TestWatch
--- FAIL: TestWatch (4.72s)
	resttest.go:1013: unexpected timeout from result channel
FAIL
FAIL	k8s.io/kubernetes/pkg/registry/controller/etcd	4.955s
```

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes https://github.com/kubernetes/kubernetes/issues/14059 (and probably more)

@lavalamp @dchen1107 @thockin

### CODE REVIEW

**Comment by quinton-hoole** [comment]
@brendandburns what is the total test execution time before and after these timeout increases?

**Comment by brendandburns** [comment]
@quinton-hoole expected case total execution time after the increases is 0.  These timeouts only trigger on failed runs, so runtime will only increase if a test fails (highly unlikely).

I'd consider making it a dynamic setting, so that we can adjust for automation (long timeout) vs test development (short timeout)

but it seems like overkill.

File: pkg/api/rest/resttest/resttest.go [hunk 1]
```diff
@@ -961,7 +961,7 @@ func (t *Tester) testWatchFields(obj runtime.Object, initWatchFn InitWatchFunc,
 				if !ok {
 					t.Errorf("watch channel should be open")
 				}
-			case <-time.After(time.Millisecond * 100):
+			case <-time.After(time.Second * 30):
```
Comment (quinton-hoole) [inline pkg/api/rest/resttest/resttest.go]: a 300x increase seems excessive?

File: pkg/api/rest/resttest/resttest.go [hunk 2]
```diff
@@ -982,7 +982,7 @@ func (t *Tester) testWatchFields(obj runtime.Object, initWatchFn InitWatchFunc,
 			select {
 			case <-watcher.ResultChan():
 				t.Errorf("unexpected result from result channel")
-			case <-time.After(time.Millisecond * 100):
+			case <-time.After(time.Millisecond * 500):
```
Comment (quinton-hoole) [inline pkg/api/rest/resttest/resttest.go]: as above 500x?

File: pkg/api/rest/resttest/resttest.go [hunk 3]
```diff
@@ -1009,7 +1009,7 @@ func (t *Tester) testWatchLabels(obj runtime.Object, initWatchFn InitWatchFunc,
 				if !ok {
 					t.Errorf("watch channel should be open")
 				}
-			case <-time.After(time.Millisecond * 100):
+			case <-time.After(time.Second * 30):
```
Comment (quinton-hoole) [inline pkg/api/rest/resttest/resttest.go]: As above.

File: pkg/api/rest/resttest/resttest.go (hunk 1 above)
Comment (quinton-hoole) [inline pkg/api/rest/resttest/resttest.go]: How long do these usually take when they pass?
I guess what I'm concerned about is that this is inside a loop (no sure how many iterations it typically does?), and the fail time for each iteration increases dramatically.  If there's a regression across all of these fields, these unit tests might take many many minutes to fail?

File: pkg/api/rest/resttest/resttest.go (hunk 2 above)
Comment (brendandburns) [inline pkg/api/rest/resttest/resttest.go]: this one's 5x actually.  100milliseconds is a slow disk seek, a GC pause, ...

File: pkg/api/rest/resttest/resttest.go (hunk 1 above)
Comment (brendandburns) [inline pkg/api/rest/resttest/resttest.go]: No, its not, I've seen this flake, and again, this only happens when the test is broken.  There's no reason at all to sleep effectively forever here.  Most of the time, the case above this timeout will trigger way before anything else, and this is just hear to catch failures.

File: pkg/api/rest/resttest/resttest.go (hunk 3 above)
Comment (brendandburns) [inline pkg/api/rest/resttest/resttest.go]: as above too ;)

**Comment by brendandburns** [comment]
@quinton-hoole comments addressed.

**Comment by quinton-hoole** [comment]
My general question is what heuristics did you use for picking the new timeouts?  Given that they're up to 500x the previous values, it would be good to document (in comments) where they came from.

**Comment by brendandburns** [comment]
The meta-point is that for any:
```go
...
<- time.After(timeout):
    t.Errorf("Timed out")
```

The timeout should effectively be "forever"  Obviously we don't want our tests to truly lock up forever, but 30s was chose to be long enough that it is effectively forever for the things that can slow down a run on a heavily contended machine (GC, seeks, etc), but not so long as to make a developer ctrl-c a test run if they do happen to break that test.

**Comment by quinton-hoole** [comment]
Fair enough. In that case I'd add a constant along the lines of "NEARLY_FOREVER" with a comment to the effect of the above, and use that consistently across all tests.  Sound reasonable?

**Comment by brendandburns** [comment]
Ok, will do

**Comment by brendandburns** [comment]
@quinton-hoole done, ptal.

Thanks
--brendan

**Comment by lavalamp** [comment]
My heuristic is that things could take at least 20x longer in CI than locally, FWIW.

File: pkg/apiserver/watch_test.go [hunk 4]
```diff
@@ -378,8 +379,8 @@ func TestWatchHTTPTimeout(t *testing.T) {
 		if !watcher.Stopped {
 			t.Errorf("Leaked watch on timeout")
 		}
-	case <-time.After(100 * time.Millisecond):
-		t.Errorf("Failed to stop watcher after 100ms of timeout signal")
+	case <-time.After(util.ForeverTestTimeout):
+		t.Errorf("Failed to stop watcher after 30s of timeout signal")
```
Comment (quinton-hoole) [inline pkg/apiserver/watch_test.go]: nit: %v sec

File: pkg/client/cache/reflector_test.go [hunk 5]
```diff
@@ -58,8 +58,8 @@ func TestCloseWatchChannelOnError(t *testing.T) {
 		if ok {
 			t.Errorf("Watch channel left open after cancellation")
 		}
-	case <-time.After(100 * time.Millisecond):
-		t.Errorf("the cancellation is at least 99 milliseconds late")
+	case <-time.After(util.ForeverTestTimeout):
+		t.Errorf("the cancellation is at least 30 seconds late")
```
Comment (quinton-hoole) [inline pkg/client/cache/reflector_test.go]: nit: %v not 30

File: pkg/client/cache/reflector_test.go [hunk 6]
```diff
@@ -87,8 +87,8 @@ func TestRunUntil(t *testing.T) {
 		if ok {
 			t.Errorf("Watch channel left open after stopping the watch")
 		}
-	case <-time.After(100 * time.Millisecond):
-		t.Errorf("the cancellation is at least 99 milliseconds late")
+	case <-time.After(util.ForeverTestTimeout):
+		t.Errorf("the cancellation is at least 30 seconds late")
```
Comment (quinton-hoole) [inline pkg/client/cache/reflector_test.go]: as above

File: test/integration/client_test.go [hunk 7]
```diff
@@ -160,8 +161,8 @@ func TestSingleWatch(t *testing.T) {
 	defer w.Stop()
 
 	select {
-	case <-time.After(5 * time.Second):
-		t.Fatal("watch took longer than 15 seconds")
+	case <-time.After(util.ForeverTestTimeout):
+		t.Fatal("watch took longer than 30 seconds")
```
Comment (quinton-hoole) [inline test/integration/client_test.go]: nit: %v not 30

**Comment by quinton-hoole** [comment]
LGTM barring minor nits.  Thanks for putting up with my pedantic review ;-)

**Comment by quinton-hoole** [comment]
PS:  Shippable claims it failed 18 tests.  Need to clear that before merging.

File: pkg/apiserver/watch_test.go (hunk 4 above)
Comment (brendandburns) [inline pkg/apiserver/watch_test.go]: done.

File: pkg/client/cache/reflector_test.go (hunk 5 above)
Comment (brendandburns) [inline pkg/client/cache/reflector_test.go]: done.

File: pkg/client/cache/reflector_test.go (hunk 6 above)
Comment (brendandburns) [inline pkg/client/cache/reflector_test.go]: done.

File: test/integration/client_test.go (hunk 7 above)
Comment (brendandburns) [inline test/integration/client_test.go]: done.

**Comment by brendandburns** [comment]
nits addressed.  unit tests pass locally, we'll see what Shippable says next.

**Comment by brendandburns** [comment]
closing and re-opening to trigger shippable...

**Comment by brendandburns** [comment]
@quinton-hoole all green ready for submit, I can haz LGTMz?

**Comment by quinton-hoole** [comment]
Yes, apologies for the delay, LGTM

**Comment by quinton-hoole** [comment]
PS:  It's not clear to my why Shippable results are not shown here?  @brendandburns ?

**Comment by brendandburns** [comment]
Closing/re-opening to ping shippable...  (sometimes it appears to miss the github hook for an update)

**Comment by sttts** [comment]
Any updates on this one? Shippable is more and more becoming a blocker.

**Comment by erictune** [comment]
@brendandburns shippable looks like a build failure.  Also, needs rebase.

**Comment by brendandburns** [comment]
Rebased...

**Comment by brendandburns** [comment]
Shippable is an instance of https://github.com/kubernetes/kubernetes/issues/13117

**Pull request merged**